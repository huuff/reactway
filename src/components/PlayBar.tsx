import { PlaybackMode } from "../game/settings";
import { faPlay, faPause, faRepeat } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { HistoryAction } from "../grid/tick-history";
import { ChangeEvent, Dispatch } from "react";
import { randomSeed } from "../util/birth-function";
import { useRouter } from "next/router";
import ClassedSlot from "./ClassedSlot";

type PlayBarProps = {
    tickDuration: number | null,
    setPlayback: (mode: PlaybackMode) => void,
    historyLength: number,
    historyPosition: number,
    dispatchTickHistory: Dispatch<HistoryAction>, // TODO: Only one option to dispatch a setPosition?
    newGame: () => void,
} & { className?: string }

// TODO: Test
const PlayBar = ({
    tickDuration,
    setPlayback,
    historyLength,
    historyPosition,
    dispatchTickHistory,
    newGame,
    className
}: PlayBarProps) => {
    function handleChange(e: ChangeEvent<HTMLInputElement>) {
        dispatchTickHistory({
            type: "setPosition",
            value: +e.currentTarget.value,
        })
    }

    return (
        <div className={`${className || ""} px-4 flex h-8`}>
            <ClassedSlot className="mr-3 h-4 text-blue-500 cursor-pointer hover:scale-125">
                {tickDuration !== null
                    ? <FontAwesomeIcon icon={faPause} onClick={() => setPlayback("pause")} />
                    : <FontAwesomeIcon icon={faPlay}  onClick={() => setPlayback("play")} />
                }
            </ClassedSlot>
            <input
                type="range"
                min="0"
                value={historyPosition}
                max={historyLength - 1}
                onChange={handleChange}
                className="grow" />

            <div className="ml-3">
                <FontAwesomeIcon className="h-4 hover:scale-125 text-green-500 cursor-pointer" icon={faRepeat} onClick={newGame} />
            </div>
        </div>
    )
}

export default PlayBar;