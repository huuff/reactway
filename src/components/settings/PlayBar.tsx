import { faPlay, faPause, faRepeat, faEraser } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Playback } from '../../settings/use-playback';
import ClassedSlot from "../util/ClassedSlot";

type PlayBarProps = {
    playback: Playback,
    historyLength: number,
    historyPosition: number,
    setHistoryPosition: (newPosition: number) => void,
    startNewGame: () => void,
    clearGrid: () => void,
} & { className?: string };

const PlayBar = ({
    playback: { isPlaying, start, pause },
    historyLength,
    historyPosition,
    setHistoryPosition,
    startNewGame,
    clearGrid,
    className
}: PlayBarProps) => {
    return (
        <div className={`${className || ""} h-8 flex items-center`}>
            <ClassedSlot className="mr-3 h-4 text-blue-500 cursor-pointer hover:scale-125">
                { isPlaying
                    ? <FontAwesomeIcon icon={faPause} data-fa-title-id="pause" onClick={pause} />
                    : <FontAwesomeIcon icon={faPlay} data-fa-title-id="play" onClick={start} />
                }
            </ClassedSlot>
            <input
                type="range"
                min="0"
                value={historyPosition}
                max={historyLength - 1}
                onChange={(e) => setHistoryPosition(+e.currentTarget.value)}
                className="grow" />

            <div className="mr-1 whitespace-nowrap">
                <FontAwesomeIcon
                    icon={faEraser}
                    data-fa-title-id="clear"
                    className="inline-block ml-2 h-4 hover:scale-125 text-red-500 cursor-pointer"
                    onClick={clearGrid} />
                <FontAwesomeIcon
                    icon={faRepeat}
                    data-fa-title-id="reset"
                    className="inline-block ml-2 h-4 hover:scale-125 text-green-500 cursor-pointer"
                    onClick={startNewGame} />
            </div>
        </div>
    );
};

export default PlayBar;
export type { PlayBarProps };