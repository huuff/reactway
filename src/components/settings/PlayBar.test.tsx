import renderer from "react-test-renderer";
import PlayBar from "./PlayBar";

describe("PlayBar", () => {
    test("matches snapshot", () => {
        expect(renderer
            .create(<PlayBar 
                playback={{ isPlaying: true, start: jest.fn(), pause: jest.fn() }}
                historyLength={10}
                historyPosition={5}
                setHistoryPosition={jest.fn()}
                startNewGame={jest.fn()}
                clearGrid={jest.fn()}
            />)
            .toJSON()).toMatchSnapshot();
    });

});