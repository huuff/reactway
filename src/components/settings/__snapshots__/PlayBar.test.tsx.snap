// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayBar matches snapshot 1`] = `
<div
  className=" h-8 flex items-center"
>
  <svg
    aria-labelledby="svg-inline--fa-title-000000000000"
    className="svg-inline--fa fa-pause  mr-3 h-4 text-blue-500 cursor-pointer hover:scale-125"
    data-icon="pause"
    data-prefix="fas"
    onClick={[MockFunction]}
    role="img"
    style={{}}
    viewBox="0 0 320 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title
      id="svg-inline--fa-title-000000000000"
      style={{}}
    >
      pause
    </title>
    <path
      d="M48 64C21.5 64 0 85.5 0 112V400c0 26.5 21.5 48 48 48H80c26.5 0 48-21.5 48-48V112c0-26.5-21.5-48-48-48H48zm192 0c-26.5 0-48 21.5-48 48V400c0 26.5 21.5 48 48 48h32c26.5 0 48-21.5 48-48V112c0-26.5-21.5-48-48-48H240z"
      fill="currentColor"
      style={{}}
    />
  </svg>
  <input
    className="grow"
    max={9}
    min="0"
    onChange={[Function]}
    type="range"
    value={5}
  />
  <div
    className="mr-1 whitespace-nowrap"
  >
    <svg
      aria-labelledby="svg-inline--fa-title-000000000000"
      className="svg-inline--fa fa-eraser inline-block ml-2 h-4 hover:scale-125 text-red-500 cursor-pointer"
      data-icon="eraser"
      data-prefix="fas"
      onClick={[MockFunction]}
      role="img"
      style={{}}
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="svg-inline--fa-title-000000000000"
        style={{}}
      >
        clear
      </title>
      <path
        d="M290.7 57.4L57.4 290.7c-25 25-25 65.5 0 90.5l80 80c12 12 28.3 18.7 45.3 18.7H288h9.4H512c17.7 0 32-14.3 32-32s-14.3-32-32-32H387.9L518.6 285.3c25-25 25-65.5 0-90.5L381.3 57.4c-25-25-65.5-25-90.5 0zM297.4 416H288l-105.4 0-80-80L227.3 211.3 364.7 348.7 297.4 416z"
        fill="currentColor"
        style={{}}
      />
    </svg>
    <svg
      aria-labelledby="svg-inline--fa-title-000000000000"
      className="svg-inline--fa fa-repeat inline-block ml-2 h-4 hover:scale-125 text-green-500 cursor-pointer"
      data-icon="repeat"
      data-prefix="fas"
      onClick={[MockFunction]}
      role="img"
      style={{}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="svg-inline--fa-title-000000000000"
        style={{}}
      >
        reset
      </title>
      <path
        d="M0 224c0 17.7 14.3 32 32 32s32-14.3 32-32c0-53 43-96 96-96H320v32c0 12.9 7.8 24.6 19.8 29.6s25.7 2.2 34.9-6.9l64-64c12.5-12.5 12.5-32.8 0-45.3l-64-64c-9.2-9.2-22.9-11.9-34.9-6.9S320 19.1 320 32V64H160C71.6 64 0 135.6 0 224zm512 64c0-17.7-14.3-32-32-32s-32 14.3-32 32c0 53-43 96-96 96H192V352c0-12.9-7.8-24.6-19.8-29.6s-25.7-2.2-34.9 6.9l-64 64c-12.5 12.5-12.5 32.8 0 45.3l64 64c9.2 9.2 22.9 11.9 34.9 6.9s19.8-16.6 19.8-29.6V448H352c88.4 0 160-71.6 160-160z"
        fill="currentColor"
        style={{}}
      />
    </svg>
  </div>
</div>
`;
